parlai/agents/rag/modules.py

```diff
@@ -0,0 +1,569 @@
+        self.n_extra_positions = opt['n_extra_positions']
+        self.n_positions = get_n_positions_from_options(opt) + opt['n_extra_positions']
+        assert opt['n_extra_positions'] >= 0
+        self.expanded_input_truncate = min(
+            opt['text_truncate'] or opt['truncate'], get_n_positions_from_options(opt)
+        )
+        if self.n_extra_positions > 0:
+            self.expanded_input_truncate = max(
+                self.expanded_input_truncate, self.n_extra_positions
+            )
+        self.min_doc_token_length = opt['min_doc_token_length']
+
+        # modules
+        self.retriever = retriever_factory(opt, dictionary, shared=retriever_shared)
+        self.seq2seq_encoder = self.build_encoder(
+            opt,
+            dictionary=dictionary,
+            embedding=self.embeddings,
+            padding_idx=self.pad_idx,
+        )
+        self.seq2seq_decoder = self.build_decoder(
+            opt, embedding=self.embeddings, padding_idx=self.pad_idx
+        )
+
+    @classmethod
+    def build_encoder(
+        cls,
+        opt: Opt,
+        *args,
+        dictionary: Optional[DictionaryAgent] = None,
+        embedding: Optional[torch.nn.Embedding] = None,
+        encoder_class: Optional[Type] = None,
+        **kwargs,
+    ):
+        if encoder_class is None:
+            assert dictionary is not None
+            return RagEncoder(
+                opt=opt, dictionary=dictionary, embedding=embedding, **kwargs
+            )
+        else:
+            return encoder_class(opt, *args, **kwargs)
+
+    @classmethod
+    def build_decoder(
+        cls,
+        opt: Opt,
+        *args,
+        embedding: Optional[torch.nn.Embedding] = None,
+        n_positions: Optional[int] = None,
+        decoder_class: Optional[Type] = None,
+        **kwargs,
+    ):
+        if decoder_class is None:
+            return RagDecoder(opt=opt, embedding=embedding, n_positions=n_positions)
+        else:
+            return decoder_class(opt, *args, **kwargs)
+
+    def tokenize_query(self, query: str) -> List[int]:
+        """
+        Tokenize the query for the retriever.
+        """
+        return self.retriever.tokenize_query(query)
+
+    def get_retriever_delimiter(self) -> str:
+        """
+        Return the retriever's delimiter.
+        """
+        return self.retriever.get_delimiter()
+
+    def encoder(
+        self,
+        input: torch.LongTensor,
+        input_lengths: torch.LongTensor,
+        query_vec: torch.LongTensor,
+        input_turns_cnt: torch.LongTensor,
+        positions: Optional[torch.LongTensor] = None,
+        segments: Optional[torch.LongTensor] = None,
+    ) -> Tuple[
+        torch.Tensor,
+        torch.BoolTensor,
+        Optional[torch.LongTensor],
+        Optional[List[List[Document]]],
+        Optional[torch.Tensor],
+    ]:
+        """
+        Expand input to seq2seq encoder via concatenating retrieved documents.
```

Comment:
nit; Maybe explicitly mention that this does the retrieving (the retrieved documents aren't passed in)

Reply:
ya good idea